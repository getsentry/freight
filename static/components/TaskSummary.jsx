var React = require('react');
var Router = require('react-router');
var Link = Router.Link;
var joinClasses = require("react/lib/joinClasses");

var Duration = require('./Duration');
var TimeSince = require('./TimeSince');

var Progress = React.createClass({
  propTypes: {
    value: React.PropTypes.number.isRequired,
  },

  render() {
    return (
      <span className="progress" style={{width: this.props.value + '%'}} />
    );
  }
});

var TaskSummary = React.createClass({
  mixins: [Router.Navigation],

  taskInProgress(task) {
    return task.status == 'in_progress' || task.status == 'pending';
  },

  getEstimatedProgress(task) {
    if (task.dateFinished) {
      return 100;
    }

    var started = new Date(task.dateStarted).getTime();
    if (!started) {
      return 0;
    }

    var now = Math.max(new Date().getTime(), started);
    return parseInt(Math.min((now - started) / 1000 / task.estimatedDuration * 100, 95), 10);
  },

  getStatusLabel(task) {
    switch (task.status) {
      case 'cancelled':
        return 'Cancelled';
      case 'failed':
        return 'Failed';
      case 'finished':
        return 'Finished';
      case 'pending':
        return 'Pending';
      case 'in_progress':
        return 'In progress';
    }
  },

  gotoTask(e) {
    if (e) {
      e.preventDefault();
    }

    this.transitionTo('deployDetails', {
      app: this.props.task.app.name,
      env: this.props.task.environment,
      number: this.props.task.number
    });
  },

  render() {
    var task = this.props.task;
    var className = 'deploy';
    if (this.taskInProgress(task)) {
      className += ' active';
    } else {
      className += ' finished';
    }
    if (task.status === 'failed') {
      className += ' failed';
    } else if (task.status === 'cancelled') {
      className += ' cancelled';
    }

    return (
      <div className={joinClasses(this.props.className, className)}
           onClick={this.gotoTask}>
        <Progress value={this.getEstimatedProgress(task)} />
        <h3>
          {task.name}
        </h3>
        <div className="ref">
          <div className="sha">{task.sha.substr(0, 7)}</div>
          {task.ref}
        </div>
        <div className="meta">
          {task.status == 'pending' &&
            <small><strong>QUEUED</strong> &mdash; </small>
          }
          {task.dateFinished ?
            <small>{this.getStatusLabel(task)} <TimeSince date={task.dateFinished} /> &mdash; <Duration seconds={task.duration} className="duration" /></small>
          : (task.dateStarted ?
            <small>Started <TimeSince date={task.dateStarted} /></small>
          :
            <small>Created <TimeSince date={task.dateCreated} /></small>
          )}
          <small> &mdash; by {task.user.name}</small>
        </div>
      </div>
    );
  }
});

export default TaskSummary;
