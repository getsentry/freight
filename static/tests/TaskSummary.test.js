import React from 'react';
import PropTypes from 'prop-types';
import {mount} from 'enzyme';

import TaskSummary from '../components/TaskSummary.jsx';

//TODO: figure out why snapshots are rendering null.
test('TaskSummary Snapshot', () => {
  const task = {
    app: {
      id: '1',
      name: 'freight',
    },
    dateCreated: '2017-07-13T18:30:29.941695Z',
    dateFinished: '2017-07-13T18:30:45.484661Z',
    dateStarted: '2017-07-13T18:30:33.481500Z',
    duration: 12,
    environment: 'production',
    estimatedDuration: 12,
    id: '797',
    name: 'freight/production',
    number: 791,
    ref: 'master',
    sha: 'af56f6bc9843956c3ce92f9e7f9abb8c196340ee',
    status: 'finished',
    user: {
      dateCreated: '2017-06-21T17:39:22.706427Z',
      id: '3',
      name: 'james.andrews@sentry.io',
    },
  };
  const context = {
    context: {
      router: {},
    },
    contextTypes: {
      router: PropTypes.object.isRequired,
    },
  };

  const wrapper = mount(<TaskSummary task={task} />, context);
  expect(wrapper).toMatchSnapshot();
});
