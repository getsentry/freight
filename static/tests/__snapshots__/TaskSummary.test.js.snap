// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskSummary Snapshot 1`] = `
<TaskSummary
  task={
    Object {
      "app": Object {
        "id": "1",
        "name": "freight",
      },
      "dateCreated": "2017-07-13T18:30:29.941695Z",
      "dateFinished": "2017-07-13T18:30:45.484661Z",
      "dateStarted": "2017-07-13T18:30:33.481500Z",
      "duration": 12,
      "environment": "production",
      "estimatedDuration": 12,
      "id": "797",
      "name": "freight/production",
      "number": 791,
      "ref": "master",
      "sha": "af56f6bc9843956c3ce92f9e7f9abb8c196340ee",
      "status": "finished",
      "user": Object {
        "dateCreated": "2017-06-21T17:39:22.706427Z",
        "id": "3",
        "name": "james.andrews@sentry.io",
      },
    }
  }
>
  <div
    className="deploy finished"
    onClick={[Function]}
  >
    <Progress
      value={100}
    >
      <span
        className="progress"
        style={
          Object {
            "width": "100%",
          }
        }
      />
    </Progress>
    <h3>
      freight/production
    </h3>
    <div
      className="ref"
    >
      <div
        className="sha"
      >
        af56f6b
      </div>
      master
    </div>
    <div
      className="meta"
    >
      <small>
        Finished
         
        <TimeSince
          date="2017-07-13T18:30:45.484661Z"
        >
          <time>
            3 years ago
          </time>
        </TimeSince>
         —
         
        <Duration
          className="duration"
          seconds={12}
        >
          <span
            className="duration"
          >
            12 sec
          </span>
        </Duration>
      </small>
      <small>
         — by 
        james.andrews@sentry.io
      </small>
    </div>
  </div>
</TaskSummary>
`;
